### DISCUSSION TITLE
propagation of "within_columns_clause" during compilation

### DISCUSSION THREAD

**Discussion by sharonyogev** [sqlalchemy/sqlalchemy#7267] [Usage Questions] [answered]
Hi,
I found something, but I'm not sure if it's a bug or I'm misusing it...
I work with SQLAlchemy 1.3.24 (hopefully to transition to 1.4 soon) and PostgreSQL 13 and encountered this issue:

I know that when I pass simple str to `order_by` or `group_by` clause it will create `_textual_label_reference`, and in compile-time, it will verify that the column really exists before I send it to the database.

For example:
```
In [1]: import sqlalchemy as sa

In [2]: from sqlalchemy.ext.declarative import declarative_base

In [3]: Base = declarative_base()

In [4]: t = sa.Table("test", Base.metadata, sa.Column("a", sa.ARRAY(sa.Integer())))

In [5]: str(sa.select([t.c.a.label("b")]).group_by("a").compile())
Out[5]: 'SELECT test.a AS b \nFROM test GROUP BY test.a'

In [6]: str(sa.select([t.c.a.label("b")]).group_by("b").compile())
Out[6]: 'SELECT test.a AS b \nFROM test GROUP BY b'

In [7]: str(sa.select([t.c.a.label("b")]).group_by("c").compile())
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
~/.venv/medigator/lib/python3.8/site-packages/sqlalchemy/sql/compiler.py in visit_textual_label_reference(self, element, within_columns_clause, **kwargs)
    834             else:
--> 835                 col = with_cols[element.element]
    836         except KeyError as ke:

KeyError: 'c'

The above exception was the direct cause of the following exception:

CompileError                              Traceback (most recent call last)
<ipython-input-9-d0bc6b091087> in <module>
----> 1 str(sa.select([t.c.a.label("b")]).group_by("c").compile())

...

CompileError: Can't resolve label reference for ORDER BY / GROUP BY / DISTINCT etc. Textual SQL expression 'c' should be explicitly declared as text('c')
```

But when I do the following query:
```
cte = sa.select([t.c.a.label("b")]).group_by("b").cte()
str(sa.select([sa.select([cte.c.b]).label('c')]).compile())
```
I also get the above exception...

I tried to debug and I think I understand it, but want to make sure...
the `label('c')` compiles the inner element with `within_columns_clause=True` since it should only search column only in the from clauses, but the inner element is a `select` and for some reason this kwargs propate down via 
```
visit_grouping -> visit_select -> visit_cte -> visit_select -> visit_clauselist -> visit_textual_label_reference
```
instead of being False when entering the `visit_cte` for example, since the CTE either way don't have access to other queries using it.

Thanks,
Sharon.

**Comment by zzzeek** [comment, marked as answer]
yes that looks like a bug, because if you print the CTE by itself, it's fine, then in context, it blows up.

you can get the query you are asking for like this, less conveniently:

```py
b = t.c.a.label("b")
cte = sa.select([b]).group_by(b).cte()
print(
    str(sa.select([sa.select([cte.c.b]).label('c')]).compile())
)
```

the order_by/group_by string thing is a convenience / intuitive kind of feature, it used to just render the string as given no matter what but this was deemed a security concern, so now it does the above lookup that seems to get confused in this specific scenario.    feel free to open a new issue.

**Reply by zzzeek** [reply]
somewhat miraculously if we just make sure that "within_columns_clause" is reset, each time we enter into a new SELECT body, so that we are again having the flag act relative to the current SELECT we are rendering, this fixes the test case given, but also everything in the test suite seems to continue passing:

```diff
diff --git a/lib/sqlalchemy/sql/compiler.py b/lib/sqlalchemy/sql/compiler.py
index 0cd568fcc6..5c3fbb2b1f 100644
--- a/lib/sqlalchemy/sql/compiler.py
+++ b/lib/sqlalchemy/sql/compiler.py
@@ -3186,6 +3186,8 @@ class SQLCompiler(Compiled):
         # passed in.  for ORM use this will convert from an ORM-state
         # SELECT to a regular "Core" SELECT.  other composed operations
         # such as computation of joins will be performed.
+        kwargs["within_columns_clause"] = False
+
         compile_state = select_stmt._compile_state_factory(
             select_stmt, self, **kwargs
         )
```

**Reply by zzzeek** [reply]
I created #7269

**Reply by sharonyogev** [reply]
Great, thanks 👍
